packages/mdc-menu-surface/foundation.ts

```diff
@@ -0,0 +1,500 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+// tslint:disable:no-bitwise object-literal-sort-keys
+
+import MDCFoundation from '@material/base/foundation';
+import {MDCMenuSurfaceAdapter} from './adapter';
+import {Corner, CornerBit, cssClasses, MenuDimensions, MenuDistance, MenuPoint, numbers, strings} from './constants';
+
+interface AutoLayoutMeasurements {
+  viewportSize: MenuDimensions;
+  viewportDistance: MenuDistance;
+  anchorSize: MenuDimensions;
+  surfaceSize: MenuDimensions;
+  bodySize: MenuDimensions;
+  windowScroll: MenuPoint;
+}
+
+class MDCMenuSurfaceFoundation extends MDCFoundation<MDCMenuSurfaceAdapter> {
+  static get cssClasses() {
+    return cssClasses;
+  }
+
+  static get strings() {
+    return strings;
+  }
+
+  static get numbers() {
+    return numbers;
+  }
+
+  static get Corner() {
+    return Corner;
+  }
+
+  /**
+   * {@see MDCMenuSurfaceAdapter} for typing information on parameters and return types.
+   */
+  static get defaultAdapter(): MDCMenuSurfaceAdapter {
+    return {
+      addClass: () => undefined,
+      removeClass: () => undefined,
+      hasClass: () => false,
+      hasAnchor: () => false,
+
+      isElementInContainer: () => false,
+      isFocused: () => false,
+      isFirstElementFocused: () => false,
+      isLastElementFocused: () => false,
+      isRtl: () => false,
+
+      getInnerDimensions: () => ({height: 0, width: 0}),
+      getAnchorDimensions: () => null,
+      getWindowDimensions: () => ({height: 0, width: 0}),
+      getBodyDimensions: () => ({height: 0, width: 0}),
+      getWindowScroll: () => ({x: 0, y: 0}),
+      setPosition: () => undefined,
+      setMaxHeight: () => undefined,
+      setTransformOrigin: () => undefined,
+
+      saveFocus: () => undefined,
+      restoreFocus: () => undefined,
+      focusFirstElement: () => undefined,
+      focusLastElement: () => undefined,
+
+      notifyClose: () => undefined,
+      notifyOpen: () => undefined,
+    };
+  }
+
+  private isOpen_: boolean;
+  private openAnimationEndTimerId_: number;
+  private closeAnimationEndTimerId_: number;
+  private animationRequestId_: number;
+  private dimensions_!: MenuDimensions; // initialized in open()
+  private anchorCorner_: Corner;
+  private anchorMargin_: MenuDistance;
+  private measurements_!: AutoLayoutMeasurements; // initialized in open()
+  private isQuickOpen_: boolean;
+  private isHoistedElement_: boolean;
+  private isFixedPosition_: boolean;
+  private position_: MenuPoint;
+
+  constructor(adapter: MDCMenuSurfaceAdapter) {
+    super(Object.assign(MDCMenuSurfaceFoundation.defaultAdapter, adapter));
+
+    this.isOpen_ = false;
```

Comment:
these can all be defined inline, since they are primitives. See lines: https://github.com/material-components/material-components-web/pull/4319/files#diff-66c8b5f3b2e270459927908be746faeeR56

Reply:
Nice! Good call. It looks like you can assign object literal values too:

```js
  private anchorCorner_: Corner = Corner.TOP_START;
  private anchorMargin_: MenuDistance = {top: 0, right: 0, bottom: 0, left: 0};
  private position_: MenuPoint = {x: 0, y: 0};
```